### PR TITLE
WIP: New plugin: i2c tools

### PR DESCRIPTION
**Pull request by Boernsman**
nymea-plugins pull request checklist:

- [x]  Make sure the pull request's title is of format "Plugin name: Add support for xyz" or "New plugin: Plugin 

- [x] Did you test the changes on hardware, if not (e.g. absence of required hardware), please mention a person to confirm it has been tested.

- [x] Did you update the plugin's README.md accordingly?

- [x] Did you update translations (`cd builddir && make lupdate`)?

- [x] If you added a new plugin, should it be added to nymea-plugins-all or nymea-plugins-maker?

### CODE REVIEW

**Comment by Boernsman** [comment]
Tested this plug-in on Disco x64 and Buster armhf

File: i2ctools/deviceplugini2ctools.cpp [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+ *  License along with this library; If not, see                           *
+ *  <http://www.gnu.org/licenses/>.                                        *
+ *                                                                         *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "deviceplugini2ctools.h"
+#include "devices/device.h"
+#include "plugininfo.h"
+
+#include <QDebug>
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: will be included with the plugininf.h and can be removed

File: i2ctools/deviceplugini2ctools.cpp [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+{
+}
+
+void DevicePluginI2cTools::discoverDevices(DeviceDiscoveryInfo *info)
+{
+    DeviceClassId deviceClassId = info->deviceClassId();
+
+    if (deviceClassId == i2cInterfaceDeviceClassId) {
+        QList<i2c_adap> busses = getI2cBusses();
+        foreach(i2c_adap adapter, busses) {
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: foreach `<space>` (

File: i2ctools/deviceplugini2ctools.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        info->finish(Device::DeviceErrorNoError);
+        return;
+    } else if (deviceClassId == i2cReadRegisterDeviceClassId) {
+
+
+        foreach (int file, m_i2cDeviceFiles) {
+            DeviceId parentDeviceId = m_i2cDeviceFiles.key(file);
+            QString interfaceName = myDevices().findById(parentDeviceId)->name();
+            QList<int> addresses = scanI2cBus(file, ScanModeAuto, m_fileFuncs.value(file), 0, 254);
+            foreach(int address, addresses) {
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: space

File: i2ctools/deviceplugini2ctools.cpp [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    info->finish(Device::DeviceErrorDeviceClassNotFound);
+}
+
+
+void DevicePluginI2cTools::setupDevice(DeviceSetupInfo *info)
+{
+    Device *device = info->device();
+    if (device->deviceClassId() == i2cInterfaceDeviceClassId) {
+        qCDebug(dcI2cTools) << "Setup i2c interface";
+
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: Here you can actually work with QFile, which makes it much easier to work with. You can find an example here: https://github.com/t-mon/nymea-sensorstation-plugin/blob/master/sensors/sht30.cpp#L69

i2c can work with `file.handle()`

File: i2ctools/deviceplugini2ctools.cpp [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        device->setStateValue(i2cInterfaceAvailableDevicesStateTypeId, state);
+
+        if(lookup_i2c_bus(device->paramValue(i2cInterfaceDeviceInterfaceParamTypeId).toByteArray()) == -1) {
+            device->setStateValue(i2cInterfaceConnectedStateTypeId, false);
+        } else {
+            device->setStateValue(i2cInterfaceConnectedStateTypeId, true);
+        }
+    } else if (device->deviceClassId() == i2cReadRegisterDeviceClassId) {
+        int file = m_i2cDeviceFiles.value(device->parentId());
+        int slaveAddress = device->paramValue(i2cReadRegisterDeviceAddressParamTypeId).toInt();
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: uint and has to be between 0x03 and 0x77

File: i2ctools/deviceplugini2ctools.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+QList<i2c_adap> DevicePluginI2cTools::getI2cBusses(void)
+{
+    struct i2c_adap *adapters;
+    int count;
+    QList<i2c_adap> i2cBusses;
+
+    adapters = gather_i2c_busses();
+    if (adapters == nullptr) {
+        qWarning(dcI2cTools()) << "Error: Out of memory!";
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: qCWarning

File: i2ctools/deviceplugini2ctools.cpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+                        !(funcs & I2C_FUNC_SMBUS_READ_BYTE))
+                    || (cmd == ScanModeQuick &&
+                        !(funcs & I2C_FUNC_SMBUS_QUICK))) {
+                continue;
+            }
+
+            /* Set slave address */
+            if (ioctl(file, I2C_SLAVE, i+j) < 0) {
+                if (errno == EBUSY) {
+                    qDebug(dcI2cTools()) << "Device found but it seems busy <UU>";
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: qCDebug

File: i2ctools/deviceplugini2ctools.cpp [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+                    continue;
+                } else {
+                    qWarning(dcI2cTools()) << "Error: Could not set address" << strerror(errno);
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: qCWarning

File: i2ctools/deviceplugini2ctools.cpp [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+                    return deviceAddresses;
+                }
+            }
+
+            /* Probe this address */
+            switch (cmd) {
+            default: /* MODE_QUICK */
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: normally you end with the default case

File: i2ctools/deviceplugini2ctools.cpp [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+QList<int> DevicePluginI2cTools::scanI2cBus(int file, ScanMode mode, unsigned long funcs, int first, int last)
+{
+    int i, j;
+    int cmd, res;
+
+    QList<int> deviceAddresses;
+
+    qDebug(dcI2cTools()) << "Start scanning i2c bus";
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: qCDebug

File: i2ctools/deviceplugini2ctools.cpp [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+    QList<i2c_adap> i2cBusses;
+
+    adapters = gather_i2c_busses();
+    if (adapters == nullptr) {
+        qWarning(dcI2cTools()) << "Error: Out of memory!";
+        return i2cBusses;
+    }
+
+    for (count = 0; adapters[count].name; count++) {
+        qDebug(dcI2cTools()) << "i2c adapter nr:" << adapters[count].nr;
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: qCDebug

File: i2ctools/deviceplugini2ctools.cpp [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+QList<i2c_adap> DevicePluginI2cTools::getI2cBusses(void)
+{
+    struct i2c_adap *adapters;
+    int count;
+    QList<i2c_adap> i2cBusses;
+
+    adapters = gather_i2c_busses();
+    if (adapters == nullptr) {
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: if (!adapters)
improves reading

File: i2ctools/deviceplugini2ctools.cpp [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+    int i, j;
+    int cmd, res;
+
+    QList<int> deviceAddresses;
+
+    qDebug(dcI2cTools()) << "Start scanning i2c bus";
+
+    for (i = 0; i < 128; i += 16) {
+        for(j = 0; j < 16; j++) {
+            fflush(stdout);
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.cpp]: for what is the stdout flush?

File: i2ctools/deviceplugini2ctools.cpp (hunk 4 above)
Comment (t-mon) [reply i2ctools/deviceplugini2ctools.cpp]: Also: you can hash the QFile class, verify if the file is still open, get better error handling if not readable etc...

File: i2ctools/deviceplugini2ctools.h [hunk 14]
```diff
@@ -20,33 +20,51 @@
...
+    Q_INTERFACES(DevicePlugin)
 
-    PairingTransactionId pairingTransactionId() const;
-    void setPairingTransactionId(const PairingTransactionId &pairingTransactionId);
+public:
+    enum ScanMode {
+        ScanModeAuto  = 0,
+        ScanModeQuick = 1,
+        ScanModeRead  = 2,
+        ScanModeFunc  =	3
```
Comment (t-mon) [inline i2ctools/deviceplugini2ctools.h]: Unneeded spaces

**Comment by mzanetti** [comment]
Needs rebasing onto current master

File: i2ctools/deviceplugini2ctools.cpp [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+        foreach (int address, addresses) {
+            state.append("0x");
+            state.append(QString::number(address, 16));
+            state.append(", ");
+        }
+        if (state.size() > 2) {
+            state.resize(state.size() - 2); //remove last colon
+        } else {
+            state = "--";
+        }
```
Comment (mzanetti) [inline i2ctools/deviceplugini2ctools.cpp]:
```
QStringList states;
foreach (int address, addresses) {
  states.append(QString("0x%1).arg(address, 2, 16));
}
QString state = states.join(", ");
if (state.isEmpty()) {
  state = "--":
}
```

File: debian/control [hunk 16]
```diff
@@ -887,6 +903,7 @@ Depends: nymea-plugin-boblight,
          nymea-plugin-serialportcommander,
          nymea-plugin-systemmonitor,
          nymea-plugin-onewire,
+         nymea-plugin-i2ctool,
```
Comment (mzanetti) [inline debian/control]: typo, s missing

**Comment by mzanetti** [comment]
According to jenkins this still fails with

deviceplugini2ctools.cpp:37:23: fatal error: i2c/smbus.h: No such file or directory

**Comment by mzanetti** [comment]
dh_missing: usr/lib/x86_64-linux-gnu/nymea/plugins/libnymea_deviceplugini2ctools.so exists in debian/tmp but is not installed to anywhere

**Comment by mzanetti** [comment]
ok, the failure on smbus.h only happens on xenial...

I'm afraid this can't land while we're still supporting xenial, at least not in this repository.

**Pull request closed without merging**